### ISSUE TITLE
warning for subclasses not annotated with `@dataclass` of superclasses that *are* annotated with `@dataclass`

### ISSUE THREAD

**Issue by netanel-haber** [astral-sh/ruff#18105]
### Summary

This situation is almost certainly a user oversight [although legal]: subclasses not annotated with `@dataclass` of superclasses that *are* annotated with `@dataclass`.

The 5 tools I checked didn't provide a diagnostic/warning for this: mypy, ruff, ty, pylint, flake8.

```python
from dataclasses import dataclass

@dataclass(kw_only=True, frozen=True)
class Base:
    x: int

# The user almost certainly forgot to annotate Child with `@dataclass(kw_only=True, frozen=True)`
class Child(Base):
    y: int  # Just a class attribute, not a dataclass field

Child(x=1,y=2)         # TypeError: Base.__init__() got an unexpected keyword argument 'y'
Child.__init__         # "Signature: Child.__init__(self, *, x: int) -> None"
```

Would it make sense for me to try to create a rule for this?
It should also be discussed whether the rule should additionally enforce that the annotations provide the same arguments (e.g. kw_only etc.) as well, but this is less important than the warning about a complete absence of the annotation as that is easy to miss across files.

Related to: https://docs.astral.sh/ruff/rules/implicit-class-var-in-dataclass/.

**Comment by ntBre** [comment]
I don't think it would make sense to try to add a _ruff_ rule for this since you'd be limited to analyzing a single file at a time. So you'd need the type inference/multi-file analysis from ty. But I don't think we're quite ready to add type-inference-powered lint rules there either.

In short, it sounds like a good idea, but I don't think it would be doable to implement it right now.

Somewhat related: https://github.com/astral-sh/ruff/issues/5837

**Comment by netanel-haber** [comment]
I assume the model for those will be similar to typescript-eslint, a ruff plugin that uses ty for type-informed linting rules.

**Comment by NeilGirdhar** [comment]
@ntBre maybe it would be worth collecting issues related to type-inference-powered (or improved) lint rules?
